sonar-go-to-slang/slang-checks/src/main/java/org/sonarsource/slang/checks/SliceEqualSimplificationCheck.java

```diff
@@ -0,0 +1,192 @@
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(FunctionDeclarationTree.class, this::visitFunctionDeclaration);
+  }
+
+  private void visitFunctionDeclaration(CheckContext ctx, FunctionDeclarationTree function) {
+    if (isSliceEqualityFunction(function)) {
+      ctx.reportIssue(function.name(), MESSAGE);
+    }
+  }
+
+  private static boolean isSliceEqualityFunction(FunctionDeclarationTree function) {
+    List<ParameterTree> params = function.formalParameters();
+    if (params.size() != 2) {
+      return false;
+    }
+
+    ParameterTree param1 = params.get(0);
+    ParameterTree param2 = params.get(1);
+
+    // Semantic checks would be needed here to verify types.
+    // For example:
+    // if (!isSlice(param1.type()) || !isSlice(param2.type())) {
+    //   return false;
+    // }
+    // if (!areElementTypesComparableAndIdentical(param1.type(), param2.type())) {
+    //   return false;
+    // }
+    // if (!isBool(function.returnType())) {
+    //   return false;
+    // }
+
+    BlockTree body = function.body();
+    if (body == null || body.statementOrExpressions().size() != 3) {
+      return false;
+    }
+
+    List<Tree> statements = body.statementOrExpressions();
+    String p1Name = param1.identifier().name();
+    String p2Name = param2.identifier().name();
+
+    return isLengthCheck(statements.get(0), p1Name, p2Name) &&
+      isElementComparisonLoop(statements.get(1), p1Name, p2Name) &&
+      isReturnTrue(statements.get(2));
+  }
+
+  private static boolean isLengthCheck(Tree statement, String p1Name, String p2Name) {
+    if (!(statement instanceof IfTree)) {
+      return false;
+    }
+    IfTree ifTree = (IfTree) statement;
+    if (ifTree.elseClause() != null || !(ifTree.condition() instanceof BinaryExpressionTree)) {
+      return false;
+    }
+    BinaryExpressionTree condition = (BinaryExpressionTree) ifTree.condition();
+    if (condition.operator() != BinaryExpressionTree.Operator.NOT_EQUAL_TO) {
+      return false;
+    }
+
+    if (!isLenCallOnParam(condition.leftOperand(), p1Name, p2Name) || !isLenCallOnParam(condition.rightOperand(), p1Name, p2Name)) {
+      return false;
+    }
+
+    BlockTree thenBlock = ifTree.thenClause();
+    return thenBlock.statementOrExpressions().size() == 1 && isReturnFalse(thenBlock.statementOrExpressions().get(0));
+  }
+
+  private static boolean isLenCallOnParam(Tree tree, String p1Name, String p2Name) {
+    if (!(tree instanceof FunctionInvocationTree)) {
+      return false;
+    }
+    FunctionInvocationTree call = (FunctionInvocationTree) tree;
+    if (!"len".equals(call.memberSelect().name()) || call.arguments().size() != 1) {
+      return false;
+    }
+    Tree arg = call.arguments().get(0);
+    if (!(arg instanceof IdentifierTree)) {
+      return false;
+    }
+    String argName = ((IdentifierTree) arg).name();
+    return argName.equals(p1Name) || argName.equals(p2Name);
```

Comment:
This can be implemented in following way:
```suggestion
    if (tree instanceof FunctionInvocationTree functionInvocation &&
        functionInvocation.memberSelect() instanceof IdentifierTree identifierTree &&
        "len".equals(identifierTree.name()) && 
        functionInvocation.arguments().size() == 1) {
      var lenArgument = functionInvocation.arguments().get(0);
      if (lenArgument instanceof IdentifierTree argId) {
        String argName = argId.name();
        return argName.equals(p1Name);
      }
    }
    return false;
```

Please try to refactor other methods to use this style of coding instead of using early return.
By leverage of `instanceof` and assignment there is less needs for manual casts.